File: dsa-problems/GameTheory/solutions/python/GMT-004-circular-nim-variant.py
from typing import List, Tuple

def circular_nim(n: int, piles: List[int]) -> str:
    memo = {}
    visiting = set()

    def solve(current_piles: Tuple[int], depth: int) -> str:
        if depth > 50:
            return "Draw"
        if current_piles in memo:
            return memo[current_piles]
        if current_piles in visiting:
            return "Draw"
        
        visiting.add(current_piles)
        
        can_reach_loss = False
        can_reach_draw = False
        has_moves = False
        
        for i in range(n):
            if current_piles[i] > 0:
                for k in range(1, current_piles[i] + 1):
                    has_moves = True
                    next_piles = list(current_piles)
                    next_piles[i] -= k
                    next_piles[(i - 1 + n) % n] += 1
                    next_piles[(i + 1) % n] += 1
                    
                    res = solve(tuple(next_piles), depth + 1)
                    
                    if res == "Second":
                        can_reach_loss = True
                        break
                    if res == "Draw":
                        can_reach_draw = True
                if can_reach_loss:
                    break
        
        visiting.remove(current_piles)
        
        if can_reach_loss:
            result = "First"
        elif not has_moves:
            result = "Second"
        elif can_reach_draw:
            result = "Draw"
        else:
            result = "Second"
            
        memo[current_piles] = result
        return result

    return solve(tuple(piles), 0)

def main():
    import sys
    # Increase recursion depth
    sys.setrecursionlimit(20000)
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    iterator = iter(data)
    try:
        n = int(next(iterator))
        piles = []
        for _ in range(n):
            piles.append(int(next(iterator)))
            
        print(circular_nim(n, piles))
    except StopIteration:
        pass

if __name__ == "__main__":
    main()

File: dsa-problems/GameTheory/solutions/java/GMT-004-circular-nim-variant.java
import java.util.*;

class Solution {
    Map<String, String> memo = new HashMap<>();
    Set<String> visiting = new HashSet<>();

    public String circularNim(int n, int[] piles) {
        return solve(n, piles, 0);
    }

    private String solve(int n, int[] piles, int depth) {
        if (depth > 50) return "Draw";
        String key = Arrays.toString(piles);
        if (memo.containsKey(key)) return memo.get(key);
        if (visiting.contains(key)) return "Draw";

        visiting.add(key);
        boolean canReachLoss = false;
        boolean canReachDraw = false;
        boolean hasMoves = false;

        for (int i = 0; i < n; i++) {
            if (piles[i] > 0) {
                for (int k = 1; k <= piles[i]; k++) {
                    hasMoves = true;
                    piles[i] -= k;
                    piles[(i - 1 + n) % n]++;
                    piles[(i + 1) % n]++;
                    
                    String res = solve(n, piles, depth + 1);
                    
                    piles[(i + 1) % n]--;
                    piles[(i - 1 + n) % n]--;
                    piles[i] += k;

                    if (res.equals("Second")) {
                        canReachLoss = true;
                        break;
                    }
                    if (res.equals("Draw")) {
                        canReachDraw = true;
                    }
                }
                if (canReachLoss) break;
            }
        }

        visiting.remove(key);
        String result;
        if (canReachLoss) result = "First";
        else if (!hasMoves) result = "Second"; // No moves -> Loss
        else if (canReachDraw) result = "Draw";
        else result = "Second"; // All moves lead to First (Win for opponent)

        memo.put(key, result);
        return result;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (sc.hasNextInt()) {
            int n = sc.nextInt();
            int[] piles = new int[n];
            for (int i = 0; i < n; i++) {
                piles[i] = sc.nextInt();
            }

            Solution solution = new Solution();
            System.out.println(solution.circularNim(n, piles));
        }
        sc.close();
    }
}
